src/scalar_mul/pippenger.rs

```diff
@@ -0,0 +1,211 @@
+    type Point = EdwardsPoint;
+
+    /// Perform variable-time, variable-base scalar multiplication using the Pippenger algorithm.
+    ///
+    /// The algorithm works as follows:
+    ///
+    /// Let `n` be a number of point-scalar pairs.
+    /// Let `w` be a window of bits (5..15, chosen based on `n`, see cost factor).
+    ///
+    /// 1. Prepare `2^(w-1) - 1` buckets with indices `[1..2^(w-1))` initialized with identity point.
+    ///    Bucket 0 is not needed as it will contain points multiplied by 0.
+    /// 2. Convert scalars to a radix-`2^w` representation with signed digits in `[-2^w/2, 2^w/2]`.
+    ///    Note: all but last digit will never be equal `2^w/2`, but that's irrelevant to us here.
+    /// 3. Starting with the last window, for each point `i=[0..n)` add it to a a bucket indexed by
+    ///    the point's scalar's value in the window.
+    /// 4. Once all points in a window are sorted into buckets, add buckets by multiplying each
+    ///    by their index. Efficient way of doing it is to start with the last bucket and compute two sums:
+    ///    intermediate sum from the last to the first, and the full sum made of all intermediate sums.
+    /// 5. Shift the resulting sum of buckets by `w` bits by using `w` doublings.
+    /// 6. Add to the return value.
+    /// 7. Repeat the loop.
+    ///
+    /// Approximate cost w/o wNAF optimizations (A = addition, D = doubling):
+    ///
+    /// ```ascii
+    /// cost = (n*A + 2*(2^w/2)*A + w*D + A)*256/w
+    ///          |          |       |     |   |
+    ///          |          |       |     |   looping over 256/w windows
+    ///          |          |       |     adding to the result
+    ///    sorting points   |       shifting the sum by w bits (to the next window, starting from last window)
+    ///    one by one       |
+    ///    into buckets     adding/subtracting all buckets
+    ///                     multiplied by their indexes
+    ///                     using a sum of intermediate sums
+    /// ```
+    ///
+    /// For large `n`, dominant factor is (n*256/w) additions.
+    /// However, if `w` is too big and `n` is not too big, then `(2^w/2)*A` could dominate.
+    /// Therefore, the optimal choice of `w` grows slowly as `n` grows.
+    ///
+    fn vartime_multiscalar_mul<I, J>(scalars: I, points: J) -> EdwardsPoint
+    where
+        I: IntoIterator,
+        I::Item: Borrow<Scalar>,
+        J: IntoIterator,
+        J::Item: Borrow<EdwardsPoint>,
+    {
+        let mut scalars = scalars.into_iter();
+        let size = scalars.by_ref().size_hint().0;
+        let w = if size < 3 {
+            1
+        } else if size < 13 {
+            2
+        } else if size < 27 {
+            3
+        } else if size < 66 {
+            4
+        } else if size < 190 {
+            5
+        } else if size < 500 {
+            6
+        } else if size < 800 {
+            7
+        } else if size < 3_000 {
+            8
+        } else if size < 6_000 {
```

Comment:
How much of a performance degradation is there by capping the window size at 8? It means that the scalar digits can fit in `i8` rather than `i16`, so they will take half the space in cache.  For instance, for `2048` scalars, we have `256/8 = 32` digits each for `64K` digits total, so using 1 byte per digit rather than 2 would save `64KB` of cache.

On the other hand, if we arrange the scalar digits properly, we may be able to achieve linear access, in which case this might not matter.

Reply:
According to an estimation using the provided ruby script, the difference is like 5-7% if you double the points from 3K->6K. It gets to 25%-35% difference when you go to 30K and 300K points respectively.

Reply:
Also, interestingly, if we reduce the support to `max(w)=8`, then we only need `w = {6,7,8}`, because w < 6 are optimal for n<190 where Straus is faster anyway. I wonder if having only three possible window sizes is beneficial in some sense?